### PR TITLE
Feature/check balance

### PR DESCRIPTION
**Pull request by david1312**
**Update LAST PUSH (DELETE file yang tidak kepake di fitur check balance)**
fungsi CheckBalance gua taruh di helpers.go guys fungsi ini returnya boolean true / false

parameter yg dilempar ada 4 (target(string), accNumber(string), amount saldo buat compare(int), db )
note :
targetnya bisa diisi "VA" atau "MAIN"
kalau va lu cek balance dari virtual acc , kalo main dari accounts

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,11 +1,10 @@
-// change the module path to your directory
```
Comment (vicimilenia) [inline go.mod:1]: no need for this comment

File: go.mod [hunk 2]
```diff
@@ -1,11 +1,10 @@
...
-module github.com/polipopoliko/tsaving/tsaving
+module github.com/david1312/tsaving
```
Comment (vicimilenia) [inline go.mod]:
```suggestion
module github.com/ndv6/tsaving
```
change to match github repository link

File: helpers/helpers.go [hunk 3]
```diff
@@ -1,11 +1,47 @@
 package helpers
 
 import (
+	"database/sql"
 	"encoding/json"
 	"net/http"
+
+	"github.com/david1312/tsaving/models"
```
Comment (vicimilenia) [inline helpers/helpers.go]:
```suggestion
	"github.com/ndv6/tsaving/models"
```
change to match github repository link

File: main.go [hunk 4]
```diff
@@ -5,8 +5,8 @@ import (
 	"log"
 	"net/http"
 
-	"github.com/polipopoliko/tsaving/tsaving/api"
-	"github.com/polipopoliko/tsaving/tsaving/database"
+	"github.com/david1312/tsaving/api"
+	"github.com/david1312/tsaving/database"
```
Comment (vicimilenia) [inline main.go]:
```suggestion
	"github.com/ndv6/tsaving/api"
	"github.com/ndv6/tsaving/database"
```
change to match github repository link

**Review by celynolivia — APPROVED** [review]
sudahh okee

File: helpers/helpers.go [hunk 5]
```diff
@@ -1,11 +1,47 @@
...
 )
 
 func HTTPError(w http.ResponseWriter, status int, errorMessage string) {
 	w.WriteHeader(status)
 	json.NewEncoder(w).Encode(map[string]string{"error": errorMessage})
 }
+
+func CheckBalance(target string, accNumber string, amount int, db *sql.DB) (status bool) {
+	if target == "MAIN" {
+		saldoasal, err := models.GetBalanceAcc(accNumber, db) // println(err.Error())
```
Comment (vicimilenia) [inline helpers/helpers.go]: change `saldoasal` to english, as per yesterday's agreement

Comment (vicimilenia) [inline helpers/helpers.go]: ```// println(err.Error())``` if not used, I think its better to delete this

File: models/virtual_accounts.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+	AccountNum string    `json:"account_num"`
+	VaBalance  int       `json:"va_balance"`
+	VaColor    string    `json:"va_color"`
+	VaLabel    string    `json:"va_label"`
+	CreatedAt  time.Time `json:"created_at"`
+	UpdateAt   time.Time `json:"updated_at"`
+}
+
+func GetBalanceVA(vaNum string, db *sql.DB) (va VirtualAccounts, err error) {
+	var balancefloat float32
```
Comment (vicimilenia) [inline models/virtual_accounts.go]: change variable names to camelCase, as per yesterday's agreement

**Review by vicimilenia — CHANGES_REQUESTED** [review]

File: models/accounts.go [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+type Accounts struct {
+	Account        int       `json:"account_id"`
+	AccountNum     string    `json:"account_num"`
+	AccountBalance int       `json:"account_balance"`
+	CreatedAt      time.Time `json:"created_at"`
+}
+
+func GetBalanceAcc(accNum string, db *sql.DB) (acc Accounts, err error) {
+	var balanceFloat float32
+	err = db.QueryRow("SELECT account_balance from accounts where account_num = ($1) ", accNum).Scan(&balanceFloat)
```
Comment (vicimilenia) [inline models/accounts.go]: add error checking here so if there's an error when running db query, the next line will not run

Comment (vicimilenia) [inline models/accounts.go]:
```suggestion
	err = db.QueryRow("SELECT account_balance FROM accounts WHERE account_num = ($1) ", accNum).Scan(&balanceFloat)
```

File: models/customers.go [hunk 8]
```diff
@@ -1 +1,26 @@
...
+	CustPassword string    `json:"cust_password"`
+	CustPict     string    `json:"cust_pict"`
+	IsVerified   string    `json:"is_verified"`
+	Channel      string    `json:"channel"`
+	CreatedAt    time.Time `json:"created_at"`
+	UpdateAt     time.Time `json:"updated_at"`
+}
+
+func GetAccountNumById(id int, db *sql.DB) (cus Customers, err error) {
+	err = db.QueryRow("SELECT account_num from customers where cust_id = ($1) ", id).Scan(&cus.AccountNum)
```
Comment (vicimilenia) [inline models/customers.go]:
```suggestion
	err = db.QueryRow("SELECT account_num FROM customers WHERE cust_id = ($1) ", id).Scan(&cus.AccountNum)
```

File: models/accounts.go (hunk 7 above)
Comment (polipopoliko) [inline models/accounts.go]: return Empty Accounts object if err != nil

**Review by vicimilenia — CHANGES_REQUESTED** [review]

File: models/accounts.go [hunk 9]
```diff
@@ -0,0 +1,20 @@
+package models
+
+import (
+	"database/sql"
+	"time"
+)
+
+type Accounts struct {
+	Account        int       `json:"account_id"`
```
Comment (polipopoliko) [inline models/accounts.go]: Please use AccountId instead

File: helpers/helpers.go [hunk 10]
```diff
@@ -1,11 +1,47 @@
...
+		saldoasal, err := models.GetBalanceVA(accNumber, db)
+		//cek dapet engga balancenya
+		println(saldoasal.VaBalance)
+		if err != nil {
+			status = false
+			return
+		}
+		//kalo balancenya dapet lanjut cek saldo yg sesuai engga sama gaboleh negatif yg diinput
+		if saldoasal.VaBalance < amount || amount <= 0 {
+			status = false
```
Comment (polipopoliko) [inline helpers/helpers.go]: No need to set the status to false since boolean default value in golang is false.

File: helpers/helpers.go [hunk 11]
```diff
@@ -1,11 +1,47 @@
...
+			return
+		}
+		status = true
+	}
+	if target == "VA" {
+		saldoasal, err := models.GetBalanceVA(accNumber, db)
+		//cek dapet engga balancenya
+		println(saldoasal.VaBalance)
+		if err != nil {
+			status = false
```
Comment (polipopoliko) [inline helpers/helpers.go]: No need to set the status to false since boolean default value in golang is false.

File: helpers/helpers.go [hunk 12]
```diff
@@ -1,11 +1,47 @@
...
+	if target == "MAIN" {
+		saldoasal, err := models.GetBalanceAcc(accNumber, db) // println(err.Error())
+		//cek dapet engga balancenya
+		if err != nil {
+			status = false
+			return
+		}
+		//kalo balancenya dapet lanjut cek saldo yg sesuai engga sama gaboleh negatif yg diinput
+		if saldoasal.AccountBalance < amount || amount <= 0 {
+			status = false
```
Comment (polipopoliko) [inline helpers/helpers.go]: No need to set the status to false since boolean default value in golang is false.

File: helpers/helpers.go [hunk 13]
```diff
@@ -1,11 +1,47 @@
...
 	w.WriteHeader(status)
 	json.NewEncoder(w).Encode(map[string]string{"error": errorMessage})
 }
+
+func CheckBalance(target string, accNumber string, amount int, db *sql.DB) (status bool) {
+	if target == "MAIN" {
+		saldoasal, err := models.GetBalanceAcc(accNumber, db) // println(err.Error())
+		//cek dapet engga balancenya
+		if err != nil {
+			status = false
```
Comment (polipopoliko) [inline helpers/helpers.go]: No need to set the status to false since boolean default value in golang is false.

File: models/customers.go [hunk 14]
```diff
@@ -1 +1,26 @@
...
+	CustName     string    `json:"cust_name"`
+	CustAddress  int       `json:"cust_address"`
+	CustPhone    string    `json:"cust_phone"`
+	CustEmail    string    `json:"cust_email"`
+	CustPassword string    `json:"cust_password"`
+	CustPict     string    `json:"cust_pict"`
+	IsVerified   string    `json:"is_verified"`
+	Channel      string    `json:"channel"`
+	CreatedAt    time.Time `json:"created_at"`
+	UpdateAt     time.Time `json:"updated_at"`
```
Comment (polipopoliko) [inline models/customers.go]: Kindly change it to UpdatedAt

File: models/customers.go [hunk 15]
```diff
@@ -1 +1,26 @@
...
+}
+
+func GetAccountNumById(id int, db *sql.DB) (cus Customers, err error) {
+	err = db.QueryRow("SELECT account_num from customers where cust_id = ($1) ", id).Scan(&cus.AccountNum)
+	return
+}
```
Comment (polipopoliko) [inline models/customers.go]: No need to be commited if the function is not used

**Review by polipopoliko — CHANGES_REQUESTED** [review]

File: models/virtual_accounts.go [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+	err = db.QueryRow("SELECT va_balance from virtual_accounts where va_num = ($1) ", vaNum).Scan(&balancefloat)
```
Comment (vicimilenia) [inline models/virtual_accounts.go]: add error checking here so if there is error when querying to database, the next line won't have to run

Comment (vicimilenia) [inline models/virtual_accounts.go]:
```suggestion
	err = db.QueryRow("SELECT va_balance FROM virtual_accounts WHERE va_num = ($1) ", vaNum).Scan(&balancefloat)
```

File: helpers/helpers.go [hunk 17]
```diff
@@ -1,11 +1,47 @@
...
+		if saldoasal.AccountBalance < amount || amount <= 0 {
+			status = false
+			return
+		}
+		status = true
+	}
+	if target == "VA" {
+		saldoasal, err := models.GetBalanceVA(accNumber, db)
+		//cek dapet engga balancenya
+		println(saldoasal.VaBalance)
```
Comment (vicimilenia) [inline helpers/helpers.go]: is it necessary to print the balance? If yes, maybe add "Your VA balance: " or something like that so user knows what number is printed on their screen

**Review by vicimilenia — CHANGES_REQUESTED** [review]

File: models/virtual_accounts.go [hunk 18]
```diff
@@ -0,0 +1,27 @@
+package models
+
+import (
+	"database/sql"
+	"time"
+)
+
+type VirtualAccounts struct {
+	VaId       int       `json:"var_id"`
```
Comment (polipopoliko) [inline models/virtual_accounts.go]: should be `json:"va_id"`

File: models/customers.go (hunk 14 above)
Comment (polipopoliko) [reply models/customers.go]: I think you forgot to change it to UpdatedAt instead of UpdateAt

File: api/virtual_accounts/virtual_accounts.go [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+	if err != nil {
+		helpers.HTTPError(w, http.StatusBadRequest, "unable to parse json request")
+		return
+	}
+
+	//cek balance
+	// fmt.Fprintf(w, "Api Called Sukses : %v", noRek.AccountNum)
+	status := helpers.CheckBalance("MAIN", vac.AccountNum, vac.VaBalance, va.db)
+	if !status {
+		helpers.HTTPError(w, http.StatusBadRequest, "saldo tidak mencukupi")
```
Comment (polipopoliko) [inline api/virtual_accounts/virtual_accounts.go]: Change the error message to "insufficient balance"

**Review by polipopoliko — CHANGES_REQUESTED** [review]
For the sake of consistency

File: api/virtual_accounts/virtual_accounts.go [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+		return
+	}
+	fmt.Fprintln(w, "lanjut coding nambahin balance ke va")
+	// json.NewEncoder(w).Encode({"status" : 1})
```
Comment (vicimilenia) [inline api/virtual_accounts/virtual_accounts.go]: if needed, please uncomment. If not, please remove/don't commit

File: api/virtual_accounts/virtual_accounts.go [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+	//CEK inputan body dari apinya dulu sesuai format json apa gak
+	var vac AddBalanceVARequest
+	err := json.NewDecoder(r.Body).Decode(&vac)
+	if err != nil {
+		helpers.HTTPError(w, http.StatusBadRequest, "unable to parse json request")
+		return
+	}
+
+	//cek balance
+	// fmt.Fprintf(w, "Api Called Sukses : %v", noRek.AccountNum)
```
Comment (vicimilenia) [inline api/virtual_accounts/virtual_accounts.go]: if needed, please uncomment. If not, please remove/don't commit

File: api/router.go [hunk 22]
```diff
@@ -6,12 +6,19 @@ import (
...
 
 	"github.com/go-chi/chi"
+	"github.com/ndv6/tsaving/api/virtual_accounts"
 )
 
 func Router(db *sql.DB) *chi.Mux {
 	chiRouter := chi.NewRouter()
 	chiRouter.Get("/", HomeHandler)
 	chiRouter.NotFound(NotFoundHandler)
+	//tambahan david
```
Comment (vicimilenia) [inline api/router.go]: if needed, please uncomment. If not, please remove/don't commit

File: api/router.go [hunk 23]
```diff
@@ -6,12 +6,19 @@ import (
...
+
+	va := virtual_accounts.NewVAHandler(db)
+	chiRouter.Post("/vac/add_balance_vac", va.AddBalanceVA)
+	chiRouter.Post("/", HomeHandler)
+	//end tambahan david
```
Comment (vicimilenia) [inline api/router.go]: if needed, please uncomment. If not, please remove/don't commit

**Review by vicimilenia — CHANGES_REQUESTED** [review]

**Comment by david1312** [comment]
Last Update **Fix few codes and delete unnecessary files **

**Pull request closed without merging**